segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,500 @@
+                                                                                                       Duration timeout) {
+        return s -> Futures.exceptionallyComposeExpecting(
+                container.forSegment(getEventProcessorSegmentName(container.getId(), s), timeout),
+                e -> e instanceof StreamSegmentNotExistsException,
+                () -> metadataStore.submitAssignment(newSegment(getEventProcessorSegmentName(container.getId(), s),
+                        SYSTEM_CRITICAL_SEGMENT, Collections.emptyList()), true, timeout) // Segment should be pinned.
+                        .thenCompose(l -> container.forSegment(getEventProcessorSegmentName(container.getId(), s), timeout)));
+    }
+
+    /**
+     * Reports the metrics periodically for each {@link ContainerEventProcessor.EventProcessor}.
+     */
+    private void reportMetrics() {
+        for (EventProcessorImpl ep : eventProcessorMap.values()) {
+            SegmentStoreMetrics.outstandingEventProcessorBytes(ep.getName(), containerId, ep.getOutstandingBytes());
+            // Only report the last iteration processing latency if it has been successful.
+            if (!ep.failedIteration.get()) {
+                ep.metrics.batchProcessingLatency(ep.lastIterationLatency.get());
+            }
+        }
+    }
+
+    //endregion
+
+    //region AutoCloseable Implementation
+
+    @Override
+    public void close() {
+        if (!closed.getAndSet(true)) {
+            log.info("{}: Closing ContainerEventProcessor service.", this.traceObjectId);
+            closeProcessors();
+            eventProcessorMetricsReporting.cancel(true);
+        }
+    }
+
+    private void closeProcessors() {
+        synchronized (eventProcessorMap) {
+            if (eventProcessorMap.isEmpty()) {
+                return;
+            }
+            eventProcessorMap.forEach((k, v) -> {
+                try {
+                    v.close();
+                } catch (Exception e) {
+                    log.warn("{}: Problem closing EventProcessor {}.", this.traceObjectId, k, e);
+                }
+            });
+            eventProcessorMap.clear();
+            log.debug("{}: Closing EventProcessors complete.", this.traceObjectId);
+        }
+    }
+
+    //endregion
+
+    //region ContainerEventProcessor Implementation
+
+    @Override
+    public CompletableFuture<EventProcessor> forConsumer(@NonNull String name,
+                                                         @NonNull Function<List<BufferView>, CompletableFuture<Void>> handler,
+                                                         @NonNull EventProcessorConfig config) {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        Preconditions.checkArgument(!name.isEmpty(), "EventProcessor name cannot be empty.");
+
+        synchronized (eventProcessorMap) {
+            // If the EventProcessor is already loaded, just return it.
+            if (eventProcessorMap.containsKey(name)) {
```

Comment:
Combine these into one. `.get` will return null if it is not there.

Reply:
Done.